src/linker/Linker.Steps/SealerStep.cs

```diff
@@ -0,0 +1,124 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Collections.Generic;
+using Mono.Cecil;
+
+namespace Mono.Linker.Steps
+{
+	public class SealerStep : BaseStep
+	{
+		public SealerStep ()
+		{
+		}
+
+		protected override void ProcessAssembly (AssemblyDefinition assembly)
+		{
+			if (Annotations.GetAction (assembly) != AssemblyAction.Link)
+				return;
+
+			if (!Context.Optimizations.IsEnabled (CodeOptimizations.Sealer, assembly))
+				return;
+
+			foreach (var type in assembly.MainModule.Types)
+				ProcessType (type);
+		}
+
+		bool IsSubclassed (TypeDefinition type)
+		{
+			foreach (var a in Context.GetAssemblies ()) {
```

Comment:
This is an [O(n^2) algorithm](https://twitter.com/BruceDawson0xB/status/1120381406700429312). Would it be better to build this map when building the caches for `GetBaseMethods`/`GetOverrides`?

Reply:
What would you cache? This is called only once for every type

Reply:
We could e.g. go over all types in all assemblies once, build a HashSet of namespace+name pairs for all types that are base types of something.

Then for each type just query the HashSet instead of iterating over everything again.

The algorithm becomes O(n) on average instead of O(n^2).

Reply:
I recently added the following to our `UnityTypeMapStep` and `UnityAnnotationStore` for a different reason.

https://gist.github.com/mrvoorhe/51c20368469666f665e46c6133b0d984

I think the API could be used to address this comment as well.

Reply:
Local cache added